File: hotalg/wish/L56.py
"""
以数组 intervals 表示若干个区间的集合，其中单个区间为 intervals[i] = [starti, endi] 。
请你合并所有重叠的区间，并返回 一个不重叠的区间数组，该数组需恰好覆盖输入中的所有区间 。
"""
class Solution(object):
    def merge(self, intervals):
        """
        :type intervals: List[List[int]]
        :rtype: List[List[int]]
        """
        if not intervals or len(intervals) == 0:
            return intervals

        ans = []
        intervals = sorted(intervals,key=lambda x:x[0])
        start = intervals[0][0]
        end = intervals[0][1]

        for interval in intervals:
            if interval[0] <= end:
                end = max(end, interval[1])
            else:
                ans.append([start, end])
                start = interval[0]
                end = interval[1]

        ans.append([start, end])
        return ans

if __name__ == '__main__':
    solu = Solution()
    print(solu.merge([[1,4],[0,4]]))

File: algorithm/src/main/java/linshen/structure/L56.java
package linshen.structure;

/*
以数组 intervals 表示若干个区间的集合，其中单个区间为 intervals[i] = [starti, endi] 。
请你合并所有重叠的区间，并返回 一个不重叠的区间数组，该数组需恰好覆盖输入中的所有区间 。
[[1,4],[5,6]]
提示：

1 <= intervals.length <= 104
intervals[i].length == 2
0 <= starti <= endi <= 104
 */

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
//todo hhhh 第一版用的差分，无法解决[1,4],[5,6]这个情形。遇到区间合并，还是优先考虑排序后遍历吧

public class L56 {
    public int[][] merge(int[][] intervals) {
        List<int[]> res = new ArrayList<>();
        int start = -1;
        int end = -1;
        Arrays.sort(intervals, Comparator.comparingInt(o -> o[0]));
        for (int[] ints : intervals) {
            int l = ints[0];
            int r = ints[1];
            if (start == -1) {
                start = l;
                end = r;
            } else {
                if (l <= end) {
                    end = Math.max(end, r);
                } else {
                    res.add(new int[]{start, end});
                    start = l;
                    end = r;
                }
            }
        }
        if (start != -1) {
            res.add(new int[]{start, end});
        }
        return res.toArray(new int[0][]);
    }
}
